### PR TITLE
feat(GH workflow): migrate periodic functional tests to GH actions

### LINKED ISSUE #10745: Migrate periodic functional tests to GH Actions
## Description

As part of the [main tracker issue](https://github.com/kubeflow/pipelines/issues/10744), we need to migrate [periodic functional tests](https://github.com/GoogleCloudPlatform/oss-test-infra/blob/master/prow/prowjobs/kubeflow/pipelines/kubeflow-pipelines-periodics.yaml) to GH Actions.

### Acceptance Criteria
* Need to be able to run locally (preferrable using KinD)
* Need to implement artifact generation ([workflow example](https://github.com/kubeflow/pipelines/blob/master/.github/workflows/backend.yml#L76), and [script example](https://github.com/kubeflow/pipelines/blob/master/scripts/deploy/github/helper-functions.sh#L252)

Impacted by this bug? Give it a 👍.

### PR DESCRIPTION
**Pull request by shruti2522**
**Description of your changes:**
Fixes #10745. I have added the workflow file and tried running it locally, but I guess there is some issue with the package versions in `requirements.txt`. 

Error Log:

```
|   × Preparing metadata (pyproject.toml) did not run successfully.
|   │ exit code: 1
|   ╰─> [40 lines of output]
|       Traceback (most recent call last):
|         File "/usr/lib/python3/dist-packages/pip/_vendor/pep517/in_process/_in_process.py", line 363, in <module>
|           main()
|         File "/usr/lib/python3/dist-packages/pip/_vendor/pep517/in_process/_in_process.py", line 345, in main
|           json_out['return_val'] = hook(**hook_input['kwargs'])
|         File "/usr/lib/python3/dist-packages/pip/_vendor/pep517/in_process/_in_process.py", line 164, in prepare_metadata_for_build_wheel
|           return hook(metadata_directory, config_settings)
|         File "/usr/lib/python3/dist-packages/setuptools/build_meta.py", line 174, in prepare_metadata_for_build_wheel
|           self.run_setup()
|         File "/usr/lib/python3/dist-packages/setuptools/build_meta.py", line 267, in run_setup
|           super(_BuildMetaLegacyBackend,
|         File "/usr/lib/python3/dist-packages/setuptools/build_meta.py", line 158, in run_setup
|           exec(compile(code, __file__, 'exec'), locals())
|         File "setup.py", line 68, in <module>
|           main()
|         File "setup.py", line 54, in main
|           setuptools.setup(
|         File "/usr/lib/python3/dist-packages/setuptools/__init__.py", line 153, in setup
|           return distutils.core.setup(**attrs)
|         File "/usr/lib/python3/dist-packages/setuptools/_distutils/core.py", line 109, in setup
|           _setup_distribution = dist = klass(attrs)
|         File "/usr/lib/python3/dist-packages/setuptools/dist.py", line 459, in __init__
|           _Distribution.__init__(
|         File "/usr/lib/python3/dist-packages/setuptools/_distutils/dist.py", line 293, in __init__
|           self.finalize_options()
|         File "/usr/lib/python3/dist-packages/setuptools/dist.py", line 837, in finalize_options
|           ep(self)
|         File "/usr/lib/python3/dist-packages/setuptools/dist.py", line 858, in _finalize_setup_keywords
|           ep.load()(self, ep.name, value)
|         File "/tmp/pip-build-env-sd7akepb/normal/local/lib/python3.10/dist-packages/cffi/setuptools_ext.py", line 216, in cffi_modules
|           add_cffi_module(dist, cffi_module)
|         File "/tmp/pip-build-env-sd7akepb/normal/local/lib/python3.10/dist-packages/cffi/setuptools_ext.py", line 49, in add_cffi_module
|           execfile(build_file_name, mod_vars)
|         File "/tmp/pip-build-env-sd7akepb/normal/local/lib/python3.10/dist-packages/cffi/setuptools_ext.py", line 25, in execfile
|           exec(code, glob, glob)
|         File "src/google_crc32c_build.py", line 47, in <module>
|           FFIBUILDER = cffi.FFI()
|         File "/tmp/pip-build-env-sd7akepb/normal/local/lib/python3.10/dist-packages/cffi/api.py", line 54, in __init__
|           raise Exception("Version mismatch: this is the 'cffi' package version %s, located in %r.  When we import the top-level '_cffi_backend' extension module, we get version %s, located in %r.  The two versions should be equal; check your installation." % (
|       Exception: Version mismatch: this is the 'cffi' package version 1.16.0, located in '/tmp/pip-build-env-sd7akepb/normal/local/lib/python3.10/dist-packages/cffi/api.py'.  When we import the top-level '_cffi_backend' extension module, we get version 1.15.0, located in '/usr/lib/python3/dist-packages/_cffi_backend.cpython-310-x86_64-linux-gnu.so'.  The two versions should be equal; check your installation.
|       [end of output]
|   
|   note: This error originates from a subprocess, and is likely not a problem with pip.
| error: metadata-generation-failed
| 
| × Encountered error while generating package metadata.
| ╰─> See above for output.
| 
| note: This is an issue with the package mentioned above, not pip.
| hint: See above for details.
[Periodic Functional Tests/run_tests]   ❌  Failure - Main Run Functional Tests
[Periodic Functional Tests/run_tests] exitcode '1': failure
[Periodic Functional Tests/run_tests] 🏁  Job failed
Error: Job 'run_tests' failed
```

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

CC @rimolive

### CODE REVIEW

File: .github/workflows/periodic.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  run_tests:
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v2
+      jobs:
+      setup_kind:
+      runs-on: ubuntu-latest
+      steps:
+    - name: Check if Docker is installed
```
Comment (rimolive) [inline .github/workflows/periodic.yml:17]: This step can be removed. GH Actions always provide docker binary

File: .github/workflows/periodic.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      run: |
+        if ! command -v docker &> /dev/null
+        then
+            echo "Docker is not installed. Installing Docker..."
+            sudo apt-get update
+            sudo apt-get install -y docker.io
+        else
+            echo "Docker is already installed."
+        fi
+    - name: Install KinD
```
Comment (rimolive) [inline .github/workflows/periodic.yml:27]: GH actions have an extension to provision KinD. Check this example: https://github.com/kubeflow/pipelines/blob/master/.github/workflows/backend.yml#L36

File: .github/workflows/periodic.yml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      run: |
+        curl -Lo kind https://kind.sigs.k8s.io/dl/v0.11.1/kind-linux-amd64
+        chmod +x kind
+        sudo mv kind /usr/local/bin
+    - name: Verify KinD installation
```
Comment (rimolive) [inline .github/workflows/periodic.yml:32]: Remove that step. Previous step if using the GH Action extension already verify installation

File: .github/workflows/periodic.yml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      run: kind version        
+    - name: Run Functional Tests
+      run: 
+        ./test/kfp-functional-test/kfp-functional-test.sh
+    - name: Collect test results
```
Comment (rimolive) [inline .github/workflows/periodic.yml:37]: Although the step is present in the workflow, it requires some changes in the bash script to generate the artifacts.
* Path creation is done [here](https://github.com/kubeflow/pipelines/blob/master/scripts/deploy/github/test-pipeline.sh#L35)
* Artifact collection is done [here](https://github.com/kubeflow/pipelines/blob/master/scripts/deploy/github/test-pipeline.sh#L49)

Basically, what you can do is redirect [this execution](https://github.com/kubeflow/pipelines/blob/master/test/kfp-functional-test/kfp-functional-test.sh#L22) to a path that you create previously.

**Review by rimolive — CHANGES_REQUESTED** [review]
Thank you for your contribution! I added some comments in the workflow code, you're doing a good job!

**Comment by shruti2522** [comment]
I have updated the code. Please take a look, @rimolive.

File: .github/workflows/periodic.yml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        cluster_name: kfp-tekton
+        kubectl_version: v1.29.2
+        version: v0.22.0
+        node_image: kindest/node:v1.29.2      
+    - name: Run Functional Tests
+      run: 
+        ./test/kfp-functional-test/kfp-functional-test.sh > periodic_tests.txt
+        mkdir -p ./test/artifacts  # Directory to store artifacts
+        cp periodic_tests.txt ./test/artifacts/
+
```
Comment (rimolive) [inline .github/workflows/periodic.yml:29]: This code is important: https://github.com/kubeflow/pipelines/blob/master/.github/workflows/backend.yml#L76-L81

File: .github/workflows/periodic.yml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    - name: Create k8s Kind Cluster
+      uses: container-tools/kind-action@v2
+      with:
+        cluster_name: kfp-tekton
+        kubectl_version: v1.29.2
+        version: v0.22.0
+        node_image: kindest/node:v1.29.2      
+    - name: Run Functional Tests
+      run: 
+        ./test/kfp-functional-test/kfp-functional-test.sh > periodic_tests.txt
```
Comment (rimolive) [inline .github/workflows/periodic.yml:26]: Let me suggest a different approach:
```
log_dir=$(mktemp -d)
./test/kfp-functional-test/kfp-functional-test.sh > $log_dir/periodic_tests.txt
```

Comment (rimolive) [reply .github/workflows/periodic.yml:26]: Why `mktemp`? It creates a random path under `/tmp` for your temporary artifact collection task. That way, the last step only needs to get all the paths created by `mktemp`, compress in a tarball, and publish to a remote path.

**Review by rimolive — CHANGES_REQUESTED** [review]
Some last changes before we start testing

File: .github/workflows/periodic.yml (hunk 6 above)
Comment (shruti2522) [reply .github/workflows/periodic.yml]: I've made the changes you suggested and included the artifact collection, @rimolive.

File: .github/workflows/periodic.yml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+      - name: Run Functional Tests
+        run: |
+          log_dir=$(mktemp -d)
+          ./test/kfp-functional-test/kfp-functional-test.sh > $log_dir/periodic_tests.txt
+      - name: Collect test results
+        if: always()
+        uses: actions/upload-artifact@v4
+        with:
+          name: periodic-functional-artifacts
+          path: $log_dir
```
Comment (rimolive) [inline .github/workflows/periodic.yml:29]: Replace this line with the same content as in https://github.com/kubeflow/pipelines/blob/master/.github/workflows/backend.yml#L81

**Review by rimolive — CHANGES_REQUESTED** [review]

File: .github/workflows/periodic.yml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+on:
+  schedule:
+    - cron: '*/5 * * * *' # Run every 5 minutes
+
+jobs:
+  run_tests:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
```
Comment (rimolive) [inline .github/workflows/periodic.yml:12]: nit: upgrade to `actions/checkout@v4`

**Comment by shruti2522** [comment]
Done @rimolive

**Comment by rimolive** [comment]
Don't forget this. https://github.com/kubeflow/pipelines/pull/10751#discussion_r1590005142

**Comment by shruti2522** [comment]
done with the changes @rimolive

File: .github/workflows/periodic.yml [hunk 9]
```diff
@@ -19,11 +19,10 @@ jobs:
           node_image: kindest/node:v1.29.2
       - name: Run Functional Tests
         run: |
-          log_dir=$(mktemp -d)
```
Comment (rimolive) [inline .github/workflows/periodic.yml]: Please, add this line back and don't make any other changes. All we need is this line.

**Review by rimolive — CHANGES_REQUESTED** [review]

File: .github/workflows/periodic.yml (hunk 9 above)
Comment (shruti2522) [reply .github/workflows/periodic.yml]: oh, might have done it by mistake. I will add it back

Comment (shruti2522) [reply .github/workflows/periodic.yml]: Done

**Comment by rimolive** [comment]
/lgtm

cc @chensun

**Review by chensun — APPROVED** [review]

**Pull request merged**